### PR TITLE
Fix statedb to differentiate empty value and delete

### LINKED ISSUE #125: Allow storing empty key in SMT
### Expected behavior
StateDB should allow inserting the empty value while possible to delete key from the SMT.

For the SMT, it is currently doing
- empty value `!vec([])` should remove the key
- empty hash should not remove the key

Therefore, there should be distinction between empty value and deleted value when passing to the SMT

### Actual behavior
However, when `StateWriter.get_updated` is called, it's returning deleted key as empty value and also empty value is empty value.
Therefore, when passing to the SMT, nothing will be deleted

### Steps to reproduce
Add test to see if the deleted key will not exist later in the stateDB

### PR DESCRIPTION
**Pull request by shuse2**
### What was the problem?

This PR resolves #125 

### How was it solved?

Update to separate empty value and deleted value

### How was it tested?

Updated unit tests in rust and added functional tests in JS

### CODE REVIEW

File: src/diff.rs [hunk 1]
```diff
@@ -1,7 +1,8 @@
 /// diff provides data structure to revert the state for StateDB.
 use crate::batch;
 use crate::codec;
-use crate::types::{Cache, KVPair, KVPairCodec, NestedVec};
+use crate::types::HashWithKind;
+use crate::types::{Cache, KVPair, KVPairCodec, NestedVec, HashKind};
```
Comment (hrmhatef) [inline src/diff.rs:5]:
```suggestion
use crate::types::{Cache, KVPair, KVPairCodec, NestedVec, HashKind, HashWithKind};
```

File: src/diff.rs [hunk 2]
```diff
@@ -80,16 +81,16 @@ impl Diff {
 
     /// revert_update returns cache value with original data.
```
Comment (hrmhatef) [inline src/diff.rs:82]:
```suggestion
    /// revert_hashed_update returns cache value with original data.
```

File: src/state/state_writer.rs [hunk 3]
```diff
@@ -194,15 +194,15 @@ impl StateWriter {
 
     /// get_updated returns all the updated key-value pairs.
```
Comment (hrmhatef) [inline src/state/state_writer.rs:195]:
```suggestion
    /// get_hashed_updated returns all the updated key-value pairs.
```

**Review by hrmhatef — CHANGES_REQUESTED** [review]

**Review by hrmhatef — APPROVED** [review]

File: src/diff.rs [hunk 4]
```diff
@@ -78,18 +78,18 @@ impl Diff {
...
-    pub fn revert_update(&self) -> Cache {
+    pub fn revert_hashed_update(&self) -> Cache {
         let mut result = Cache::new();
         for kv in self.updated.iter() {
-            result.insert(kv.key_as_vec(), kv.value_as_vec());
+            result.insert(kv.key_as_vec().hash_with_kind(HashKind::Key), kv.value_as_vec().hash_with_kind(HashKind::Value));
         }
         for kv in self.deleted.iter() {
-            result.insert(kv.key_as_vec(), kv.value_as_vec());
+            result.insert(kv.key_as_vec().hash_with_kind(HashKind::Key), kv.value_as_vec().hash_with_kind(HashKind::Value));
```
Comment (ricott1) [inline src/diff.rs:89]: Here `kv.value_as_vec` is always equal to empty vec?

**Review by ricott1 — APPROVED** [review]

**Pull request merged**